enocean/prepare_packet_data.py

```diff
@@ -0,0 +1,390 @@
+#!/usr/bin/env python3
+# vim: set encoding=utf-8 tabstop=4 softtabstop=4 shiftwidth=4 expandtab
+#########################################################################
+#  Copyright 2018 Frank Häfele                       mail@frankhaefele.de
+#########################################################################
+#  Prepare Packet Data of Enocean plugin for SmartHomeNG.
+#  https://github.com/smarthomeNG//
+#
+#  This plugin is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This plugin is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this plugin. If not, see <http://www.gnu.org/licenses/>.
+#########################################################################
+import logging
+#import time
+from lib.utils import Utils
+from lib.model.smartplugin import SmartPlugin
+
+
+
+class Prepare_Packet_Data(SmartPlugin):
+    ALLOW_MULTIINSTANCE = False
+    PLUGIN_VERSION = "1.3.3"
```

Comment:
Is this really a plugin?

Reply:
No you are totally right it is not a Plugin, but as I introduced the new file I got always a failure in the logger. I will check this issue again.

Reply:
When I comment out the line 30 and line 31 I get the following failure:

2018-02-08  21:04:44 ERROR    Main         Plugin 'enocean' from section 'enocean' exception: 'Plugin' subclasses should have a 'PLUGIN_VERSION' attribute
Traceback (most recent call last):
  File "/usr/local/smarthome/lib/plugin.py", line 111, in __init__
    plugin_thread = PluginWrapper(smarthome, plugin, classname, classpath, args, instance, self.meta)
  File "/usr/local/smarthome/lib/plugin.py", line 413, in __init__
    exec("self.plugin.__init__(smarthome{0}{1})".format("," if len(arglist) else "", argstring))
  File "<string>", line 1, in <module>
  File "/usr/local/smarthome/plugins/enocean/__init__.py", line 188, in __init__
    self.prepare_packet_data = prepare_packet_data.Prepare_Packet_Data()
  File "/usr/local/smarthome/lib/model/smartplugin.py", line 348, in __new__
    raise NotImplementedError("'Plugin' subclasses should have a 'PLUGIN_VERSION' attribute")
NotImplementedError: 'Plugin' subclasses should have a 'PLUGIN_VERSION' attribute
2018-02-08  21:04:44 INFO     Main         Load of plugins finished

What to do?

Reply:
If I remove the "Smartplugin"
--> class Prepare_Packet_Data():

I get the error when I wanto to use the has_iattr()

Traceback (most recent call last):
  File "/usr/local/smarthome/lib/item.py", line 824, in __update
    method(self, caller, source, dest)
  File "/usr/local/smarthome/plugins/enocean/__init__.py", line 523, in update_item
    id_offset, rorg, payload, optional = self.prepare_packet_data.PrepareData(item, tx_eep)
  File "/usr/local/smarthome/plugins/enocean/prepare_packet_data.py", line 57, in PrepareData
    if self.has_iattr(item.conf, 'enocean_tx_id_offset'):
AttributeError: 'Prepare_Packet_Data' object has no attribute 'has_iattr'

Do you have a solution?

The point is I wanna use the function "has_iattr"

Reply:
Yes, I saw this already in your code. When the `has_iattr()` (and other methods) were introduced I never understood why this is part of `SmartPlugin`. I would introduce a new configuration objects which handles the 'has_iattr()' (and other stuff) transparently to the developer. This configuration object could be used instead of all of these conf-related methods. But for the moment we do not have this implementation and we need to use these methods directly.

One idea would be to inject all the required parameters to the method instead of using the `Item` object. This could lead to a large parameter list.

Another idea is to write a wrapper object which gets all the required parameters and which is passed to the methods.

Reply:
Hi ohinckel with this import header it works...
---
import logging
from lib.utils import Utils
from lib.model.smartplugin import SmartPlugin as Tools

class Prepare_Packet_Data(Tools):
    PLUGIN_VERSION = "1.3.3"
---
The Version info is needed but not the Multiinstance
Should we maybe ask msinn if he has an idea?